From 6e50fd5b231dcccede3a7874594b6cf921a78f6a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Pierret=20=28fepitre=29?=
 <frederic.pierret@qubes-os.org>
Date: Sat, 31 Oct 2020 14:57:49 +0100
Subject: [PATCH] Define build dates/time based on SOURCE_DATE_EPOCH

Varibles impacting reproducibility are now defined with respect
to SOURCE_DATE_EPOCH.

Improvements from Jan Beuluch <jbeulich@suse.com>.
---
 Config.mk                         | 2 ++
 INSTALL                           | 6 ++++++
 tools/firmware/hvmloader/Makefile | 2 +-
 tools/firmware/vgabios/Makefile   | 3 ++-
 xen/Makefile                      | 5 +++--
 5 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/Config.mk b/Config.mk
index 259e68863d..ab8a232693 100644
--- a/Config.mk
+++ b/Config.mk
@@ -269,3 +269,5 @@ QEMU_UPSTREAM_LOC ?= $(call or,$(wildcard $(QEMU_UPSTREAM_INTREE)),\
                                $(QEMU_UPSTREAM_URL))
 
 CONFIG_TESTS       ?= y
+
+DATE_EPOCH_OPTS := $(if $(SOURCE_DATE_EPOCH),-u -d "@$(SOURCE_DATE_EPOCH)")
\ No newline at end of file
diff --git a/INSTALL b/INSTALL
index 0d3eb89f02..2ce1f78b4d 100644
--- a/INSTALL
+++ b/INSTALL
@@ -241,6 +241,12 @@ XEN_BUILD_TIME=hh:mm:ss
 SMBIOS_REL_DATE=mm/dd/yyyy
 VGABIOS_REL_DATE="dd Mon yyyy"
 
+It is sufficient to define only SOURCE_DATE_EPOCH instead of the above
+variables for having reproducible builds. By default, they use
+SOURCE_DATE_EPOCH if defined. For specifications of SOURCE_DATE_EPOCH,
+we refer to https://reproducible-builds.org/docs/source-date-epoch.
+SOURCE_DATE_EPOCH=
+
 This variable can be used to point to a different git binary to be used.
 GIT=
 
diff --git a/tools/firmware/hvmloader/Makefile b/tools/firmware/hvmloader/Makefile
index e980ce7c5f..6b6233cf51 100644
--- a/tools/firmware/hvmloader/Makefile
+++ b/tools/firmware/hvmloader/Makefile
@@ -21,7 +21,7 @@ XEN_ROOT = $(CURDIR)/../../..
 include $(XEN_ROOT)/tools/firmware/Rules.mk
 
 # SMBIOS spec requires format mm/dd/yyyy
-SMBIOS_REL_DATE ?= $(shell date +%m/%d/%Y)
+SMBIOS_REL_DATE ?= $(shell date $(DATE_EPOCH_OPTS) +%m/%d/%Y)
 
 CFLAGS += $(CFLAGS_xeninclude)
 
diff --git a/tools/firmware/vgabios/Makefile b/tools/firmware/vgabios/Makefile
index 3284812fde..4350ef4021 100644
--- a/tools/firmware/vgabios/Makefile
+++ b/tools/firmware/vgabios/Makefile
@@ -5,7 +5,8 @@ BCC = bcc
 AS86 = as86
 
 RELEASE = `pwd | sed "s-.*/--"`
-VGABIOS_REL_DATE ?= `date '+%d %b %Y'`
+VGABIOS_REL_DATE ?= $(shell date $(DATE_EPOCH_OPTS) +%d %b %Y)
+
 RELVERS = `pwd | sed "s-.*/--" | sed "s/vgabios//" | sed "s/-//"`
 
 VGABIOS_DATE = "-DVGABIOS_DATE=\"$(VGABIOS_REL_DATE)\""
diff --git a/xen/Makefile b/xen/Makefile
index bf0c804d43..3820bf1590 100644
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -8,8 +8,9 @@ export XEN_FULLVERSION   = $(XEN_VERSION).$(XEN_SUBVERSION)$(XEN_EXTRAVERSION)
 
 export XEN_WHOAMI	?= $(USER)
 export XEN_DOMAIN	?= $(shell ([ -x /bin/dnsdomainname ] && /bin/dnsdomainname) || ([ -x /bin/domainname ] && /bin/domainname || echo [unknown]))
-export XEN_BUILD_DATE	?= $(shell LC_ALL=C date)
-export XEN_BUILD_TIME	?= $(shell LC_ALL=C date +%T)
+
+export XEN_BUILD_DATE	?= $(shell LC_ALL=C date $(DATE_EPOCH_OPTS))
+export XEN_BUILD_TIME	?= $(shell LC_ALL=C date $(DATE_EPOCH_OPTS) +%T)
 export XEN_BUILD_HOST	?= $(shell hostname)
 
 # Best effort attempt to find a python interpreter, defaulting to Python 3 if
-- 
2.26.2

